libp2p/network/url_conversion.go

```diff
@@ -0,0 +1,107 @@
+package network
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+
+	"github.com/multiformats/go-multiaddr"
+)
+
+const (
+	// anyScheme is a string which fills the "scheme" degree of freedom for
+	// a URL. Used to pad URLs without schemes such that they can be parsed
+	// via stdlib URL parser.
+	anyScheme = "scheme://"
+)
+
+// Libp2pMultiaddrFromServiceUrl transforms a URL into its libp2p multiaddr equivalent.
+// (see: https://github.com/libp2p/specs/blob/master/addressing/README.md#multiaddr-basics)
+func Libp2pMultiaddrFromServiceUrl(serviceUrl string) (multiaddr.Multiaddr, error) {
+	peerUrl, err := url.Parse(anyScheme + serviceUrl)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"parsing peer service URL: %s: %w",
+			serviceUrl,
+			err,
+		)
+	}
+
+	var (
+		// TECHDEBT: assuming TCP; remote peer's transport type must be knowable!
+		// (ubiquitously switching to multiaddr, instead of a URL, would resolve this)
+		peerTransportStr = "tcp"
+		peerIPVersionStr = "ip4"
+	)
+
+	// TECHDEBT: there's probably a more conventional way to do this.
+	// Check if we're dealing with IPv4 or IPv6
+	if strings.Count(peerUrl.Hostname(), ":") > 0 {
```

Comment:
Seems like we can just use the .net package and convert to Ip4. If it's nil, it's ip6 https://pkg.go.dev/net#IP.To4

Reply:
Nice one!